Wangscape/tilegen/CartesianPowerIterator.h

```diff
@@ -0,0 +1,161 @@
+#include "common.h"
+
+namespace tilegen
+{
+template<typename InputIt>
+class CartesianPowerIterator : public std::iterator<std::input_iterator_tag,
+                                                    std::vector<typename InputIt::value_type>>
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag,
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
+    typedef std::vector<typename InputIt::value_type> Values;
+    typedef std::vector<InputIt> Iterators;
+
+    CartesianPowerIterator(InputIt first_, InputIt last_, InputIt init, size_t power);
+    CartesianPowerIterator& operator++();
+
+    bool operator==(const CartesianPowerIterator& other) const;
+    bool operator!=(const CartesianPowerIterator& other) const;
+
+    const Values& operator*() const;
+    const typename InputIt::value_type& get(size_t n) const;
+    const Iterators& getIterators() const;
+    const InputIt& getIterator(size_t n) const;
+    const typename InputIt::difference_type coordinate(size_t n) const;
+    const InputIt& getFirst() const;
+    const InputIt& getLast() const;
+    std::pair<size_t, size_t> coordinates_2d() const;
+private:
+    const InputIt mFirst;
+    const InputIt mLast;
+    std::vector<InputIt> mIterators;
+    std::vector<typename InputIt::value_type> mValues;
+};
+
+template<typename InputIt>
+inline CartesianPowerIterator<InputIt>::CartesianPowerIterator(InputIt first, InputIt last, InputIt init, size_t power) :
+    mFirst(first), mLast(last),
+    mIterators(power, init)
+{
+    if (init != last)
+        for (const auto& it : mIterators)
+            mValues.push_back(*it);
+}
+
+template<typename InputIt>
+inline CartesianPowerIterator<InputIt>& CartesianPowerIterator<typename InputIt>::operator++()
+{
+    for (size_t i = 0; i < mIterators.size(); i++)
+    {
+        ++mIterators[i];
+        if (i < mIterators.size() - 1)
+        {
+            if (mIterators[i] == getLast())
+            {
+                mIterators[i] = getFirst();
+                mValues[i] = *mIterators[i];
+            }
+            else
+            {
+                mValues[i] = *mIterators[i];
+                break;
+            }
+        }
+        else if (mIterators[i] != getLast())
+        {
+            mValues[i] = *mIterators[i];
+        }
+    }
+    return *this;
+}
+
+template<typename InputIt>
+inline bool CartesianPowerIterator<InputIt>::operator==(const CartesianPowerIterator& other) const
+{
+    if (getFirst() != other.getFirst())
+        return false; // different ranges
+    if (getLast() != other.getLast())
+        return false; // different ranges
+    if (getIterators().size() != other.getIterators().size())
+        return false; // different powers
+    if (*getIterators().crbegin() == getLast() &&
+        *other.getIterators().crbegin() == getLast())
+        return true; // both are cend
```

Comment:
The "both are `cend`" comment is necessary because it's the only line of code that clearly defines what it means for a `CartesianPowerIterator` to be equal to `cend()`. If I move that check into a function `isEnd()` then all these comments can be removed.